src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs

```diff
@@ -431,35 +437,76 @@ public void TestLoadedGeneratorOrderIsDeterministic()
             var allGenerators = reference.GetGeneratorsForAllLanguages().Select(g => g.GetType().FullName).ToArray();
-            Assert.Equal(10, allGenerators.Length);
+            Assert.Equal(12, allGenerators.Length);
             Assert.Equal("Microsoft.CodeAnalysis.UnitTests.AnalyzerFileReferenceTests+SomeType+NestedGenerator", allGenerators[0]);
             Assert.Equal("Microsoft.CodeAnalysis.UnitTests.AnalyzerFileReferenceTests+TestGenerator", allGenerators[1]);
             Assert.Equal("Microsoft.CodeAnalysis.UnitTests.BaseGenerator", allGenerators[2]);
             Assert.Equal("Microsoft.CodeAnalysis.UnitTests.CSharpAndVisualBasicGenerator", allGenerators[3]);
             Assert.Equal("Microsoft.CodeAnalysis.UnitTests.ExplicitCSharpOnlyGenerator", allGenerators[4]);
             Assert.Equal("Microsoft.CodeAnalysis.UnitTests.SubClassedGenerator", allGenerators[5]);
             Assert.Equal("Microsoft.CodeAnalysis.UnitTests.TestGenerator", allGenerators[6]);
-            Assert.Equal("Microsoft.CodeAnalysis.UnitTests.VisualBasicAndCSharpGenerator", allGenerators[7]);
-            Assert.Equal("Microsoft.CodeAnalysis.UnitTests.FSharpGenerator", allGenerators[8]);
-            Assert.Equal("Microsoft.CodeAnalysis.UnitTests.VisualBasicOnlyGenerator", allGenerators[9]);
+            Assert.Equal("Microsoft.CodeAnalysis.IncrementalToSourceGeneratorWrapper`1[[Microsoft.CodeAnalysis.UnitTests.TestIncrementalGenerator, Microsoft.CodeAnalysis.UnitTests, Version=42.42.42.42, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]", allGenerators[7]);
+            Assert.Equal("Microsoft.CodeAnalysis.UnitTests.TestSourceAndIncrementalGenerator", allGenerators[8]);
+            Assert.Equal("Microsoft.CodeAnalysis.UnitTests.VisualBasicAndCSharpGenerator", allGenerators[9]);
+            Assert.Equal("Microsoft.CodeAnalysis.UnitTests.FSharpGenerator", allGenerators[10]);
+            Assert.Equal("Microsoft.CodeAnalysis.UnitTests.VisualBasicOnlyGenerator", allGenerators[11]);
+        }
+
+        [Fact]
+        public void TestUnwrapType()
+        {
+            AnalyzerFileReference reference = CreateAnalyzerFileReference(Assembly.GetExecutingAssembly().Location);
+
+            // analyzers (never wrapped)
+            foreach (var analyzer in reference.GetAnalyzersForAllLanguages())
+            {
+                var analyzerType = analyzer.GetType();
+                var unwrapped = AnalyzerFileReference.UnwrapTypeIfRequired(analyzerType);
+
+                Assert.Equal(analyzerType, unwrapped);
+            }
+
+            // generators
+            foreach (var generator in reference.GetGeneratorsForAllLanguages())
+            {
+                var generatorType = generator.GetType();
+                var unwrapped = AnalyzerFileReference.UnwrapTypeIfRequired(generatorType);
+                bool isWrapped = generatorType.IsGenericType && generatorType.GetGenericTypeDefinition() == typeof(IncrementalToSourceGeneratorWrapper<>);
```

Comment:
nit: this level of logic in the test sort of consititutes reimplementing the method under test within the test. Consider hard-coding the names of generators for which 'isWrapped' should be 'true'.